### PR TITLE
Driver keysight dmm time trace

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
Add a `timetrace` parameter similar to the Keithley's recently added time trace parameter (#1684 ).

Changes proposed in this pull request:
- ~Allow instrument configuration to change~ (kicked off into seperate PR)
- Fix the dreaded `volt` parameter
- Add time trace

~For this PR to properly work, #1749 should be merged (a few of its commits live on this branch until then). Then this branch should be cleaned of the overlapping commits and rebased on master.~ Okay, that has happened now 🙂

### CODE REVIEW

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py [hunk 1]
```diff
@@ -539,6 +692,8 @@ def init_measurement(self) -> None:
 
     def reset(self) -> None:
         self.write('*RST')
+        self.ask('*OPC?')
+        self.snapshot(update=True)
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Could you clarify why this change is needed? What does opc call and snapshot call in this case?

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py [hunk 2]
```diff
@@ -514,6 +637,36 @@ def __init__(self, name: str, address: str, silent: bool=False,
...
+        self.add_parameter('time_axis',
+                           label='Time',
+                           unit='s',
+                           snapshot_value=False,
+                           vals=vals.Arrays(shape=(self.timetrace_npts,)),
+                           parameter_class=TimeAxis)
+
+        self.add_parameter('timetrace',
+                           vals=vals.Arrays(shape=(self.timetrace_npts,)),
+                           setpoints=(self.time_axis,),
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]:
```suggestion
                           setpoints=(self.time_axis,),
                           snapshot_value=False,
```

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py [hunk 3]
```diff
@@ -282,6 +287,111 @@ def clear(self) -> None:
...
+
+    def _acquire_time_trace(self) -> np.ndarray:
+        """
+        The function that prepares the measurement and fetches the data
+        """
+        dt = self.instrument.timetrace_dt()
+        npts = self.instrument.timetrace_npts()
+        meas_time = npts*dt
+        disp_text = f"Acquiring {npts} samples"  # display limit: 40 characters
+        new_timeout = max(1.25 * meas_time, self.instrument.timeout())
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Please make a private attribute for the magic value of 1.25 (time trace timeout factor or smth)

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py [hunk 4]
```diff
@@ -282,6 +287,111 @@ def clear(self) -> None:
...
+        if self.instrument.has_DIG:
+            param_settings.append((self.instrument.sample.pretrigger_count,
+                                   0))
+
+        with ExitStack() as stack:
+            for ps in param_settings:
+                stack.enter_context(ps[0].set_to(ps[1]))
+
+            self.instrument.init_measurement()
+            self.instrument.write("*TRG")
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Theres a method for this :)
```suggestion
            self.instrument.trigger.force()
```

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py [hunk 5]
```diff
@@ -282,6 +287,111 @@ def clear(self) -> None:
...
+        The function that prepares the measurement and fetches the data
+        """
+        dt = self.instrument.timetrace_dt()
+        npts = self.instrument.timetrace_npts()
+        meas_time = npts*dt
+        disp_text = f"Acquiring {npts} samples"  # display limit: 40 characters
+        new_timeout = max(1.25 * meas_time, self.instrument.timeout())
+
+        param_settings = [(self.instrument.trigger.count, 1),
+                          (self.instrument.trigger.source, "BUS"),
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Do we really want to limit the use of this parameter to self triggering use case? If so, i don't mind, but this has to be documented in the code and in the example notebook.

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py (hunk 1 above)
Comment (WilliamHPNielsen) [reply qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: We always want the snapshot to stay up to date with the parameters. Calling `reset` potentially changes all values, so we better update our beliefs of their values.

`*OPC?` ensures that the instrument is done resetting before we start querying the parameters for their values. Without that line, we get the old values.

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py (hunk 2 above)
Comment (WilliamHPNielsen) [reply qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Happens automatically for a `ParameterWithSetpoint`. For reference:
https://github.com/QCoDeS/Qcodes/blob/9bf9494861897198660255ac69ff7bb498d8f9b0/qcodes/instrument/parameter.py#L1060-L1065

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py (hunk 1 above)
Comment (astafan8) [reply qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: sweet. thank you!

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py (hunk 3 above)
Comment (WilliamHPNielsen) [reply qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Good point.

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py (hunk 4 above)
Comment (WilliamHPNielsen) [reply qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: Noice!

File: qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py (hunk 5 above)
Comment (WilliamHPNielsen) [reply qcodes/instrument_drivers/Keysight/private/Keysight_344xxA_submodules.py]: I'm not completely sure what you mean. Let me know if you think the notebook sufficiently documents this fact.
